Adds illegal anonymous function error msg

ruiconti opened the pull request:
The only legal scenario for a React component function to be anonymous i.e. unnamed, is when it is passed as a callback to either `React.forwardRef` or `React.memo`.

This can be verified on [eslint's RulesOfHooks](https://github.com/facebook/react/blob/51947a14bb24bd151f76f6fc0acdbbc404de13f7/packages/eslint-plugin-react-hooks/). It uses a different eliminatory logic to rule out violations than this project, so we won't have a 1:1 match in terms of rule verification. But essentially, if a hook's ancestor (AST) node —the function declaration statement— is not in a forwardRef or memo call AND has no valid identifier (conditional stated [here](https://github.com/facebook/react/blob/51947a14bb24bd151f76f6fc0acdbbc404de13f7/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L351)), than it falls back to [this branch](https://github.com/facebook/react/blob/51947a14bb24bd151f76f6fc0acdbbc404de13f7/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L478). 

Current behavior would treat this error as "A hook hook cannot be used inside of another function", which is misleading. 

Resolves #29.

ruiconti committed:
Adds rule that verifies if anonymous function is properly used

ruiconti committed:
Adds tiny set of tests

Gelio commented on src/react-hooks-nesting-walker/error-messages.ts:
```diff
@@ -17,4 +17,6 @@ export const ERROR_MESSAGES = {
   invalidFunctionExpression: 'A hook cannot be used inside of another function',
   hookAfterEarlyReturn: 'A hook should not appear after a return statement',
+  anonymousFunctionIllegalCallback:
+    'Hook is in an anonymous function that is passed to an illegal callback',
```
I wonder if it woulld be more helpful if we provided more information on this violation, explaining what an illegal callback is. This would match [the behavior of the ESLint rule](https://github.com/facebook/react/blob/51947a14bb24bd151f76f6fc0acdbbc404de13f7/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L482-L485).

I know the existing messages are pretty terse, but your point made in #29 about providing more helpful error messages made me think if having that extra bit of info would be helpful to users of this TSLint plugin

ruiconti replied:
I agree. Though I'm not familiar on how to properly format (new line, expected format, etc) and test that string as it got a bit wide.

Please let me know what do you think of it.

Gelio commented on src/react-hooks-nesting-walker/react-hooks-nesting-walker.ts:
```diff
@@ -180,6 +180,21 @@ export class ReactHooksNestingWalker extends RuleWalker {
+       * Detect if the unnamed expression is wrapped in a illegal function call
+       */
+      if (
+        isCallExpression(ancestor.parent) &&
+        isIdentifier(ancestor.parent.expression) &&
+        !isReactComponentDecorator(ancestor.parent.expression)
```
Looking at this logic, it looks similar to the condition in the next `if` below. Do you think we could refactor the code to avoid making the same checks twice?

Something along the lines of:

```ts
if (isCallExpression(ancestor.parent)) {
  if (!isReactComponentRecorator(ancestor.parent.expression)) {
    // failure
  }

  return;
}
```

ruiconti replied:
I actually don't like making the code complex by coupling conditionals w/ conditionals above. Precisely because we'd be trading simplicity and readability for a couple of expressions. But that code can be simplified further.

Gelio commented on src/react-hooks-nesting-walker/react-hooks-nesting-walker.ts:
```diff
@@ -180,6 +180,21 @@ export class ReactHooksNestingWalker extends RuleWalker {
+       */
+      if (
+        isCallExpression(ancestor.parent) &&
+        isIdentifier(ancestor.parent.expression) &&
```
I'm wondering if this extra `isIdentifier` check is meaningful. I'm not saying this is wrong by any means. I'm just wondering if this case is different from the `if` below, where we check `isCallExpression` and `isReactComponentDecorator`, but not `isIdentifier` :thinking:

ruiconti replied:
It is. In fact, it is the only meaningful check that we need to do, given that we've reached that expression. All the others have been ruled out and returned early.

Gelio commented on test/tslint-rule/anonymous-function-error.ts.lint:
```diff
@@ -0,0 +1,19 @@
+const IllegalComponent = wrapper(function() {
+  useEffect(() => {});
+  ~~~~~~~~~~~~~~~~~~~ [Hook is in an anonymous function that is passed to an illegal callback]
```
Nice, this looks to be working well :+1: Thanks for adding the tests!

Gelio commented on test/tslint-rule/anonymous-function-error.ts.lint:
```diff
@@ -0,0 +1,19 @@
+const MemoizedComponent = React.memo((props) => {
+  const [state, setState] = React.useState(props.initValue);
+  return <span>{state}</span>
+})
```
Nit:

Looks like there's no trailing newline in this file. Could you add it to match the formatting of other files?

Gelio reviewed (commented):
Thank you for your contribution! I gave it a quick read and it looks to be working well :+1: I appreciate the thorough PR description, links to exact code from the ESLint rule help a lot

I left some comments and questions to broaden my understanding and hopefully make it even better.

I will have time probably next week to give it a deeper look and also fix the Azure Pipelines, which seem not to be working

Gelio commented on src/react-hooks-nesting-walker/error-messages.ts:
```diff
@@ -17,4 +17,6 @@ export const ERROR_MESSAGES = {
     'A hook cannot be used inside of another function',
   invalidFunctionExpression: 'A hook cannot be used inside of another function',
```
Actually, I've had a thought - wouldn't this PR be as simple as just modifying this message to make it more complete?

I didn't investigate it deeply, but this looks like the error message that is reported exactly in the situations we're trying to improve. Let me know if I'm mistaken

ruiconti replied:
That is a good point. But I don't think so. There is one illegal scenario where a hook is instantiated within an anonymous function that within an anonymous legal component function. For that we need the "A hook cannot be used inside of another function" error message.

I added one test scenario to cover for it.

Gelio replied:
Ok, I think I know what you mean now :smile: An equivalent of [`isSomewhereInsideComponentOrHook` from the ESLint rule](https://github.com/facebook/react/blob/51947a14bb24bd151f76f6fc0acdbbc404de13f7/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L500)

To implement that, we'd probably need to keep track of the surrounding context (all ancestor functions for a given hook call). For example, for:

```ts
const Functor = function() {
  const cb = useCallback(() => {
    const r = React.useRef()
  }, [])
}
```

for the innermost `useRef`, the function stack is:

1. The anonymous function expression assigned to `Functor`
2. The arrow function passed to `useCallback`

The top of the stack would tell us if it's a valid hook call or not. If we wanted to get the `isSomewhereInsideComponentOrHook`, we'd need to check if there's a valid component/hook declaration somewhere on that stack. In this case, there is (1), so we could show the `invalidFunctionExpression` warning, and if there wasn't, we'd show `anonymousFunctionIllegalCallback`.

Not sure if that's robust enough to handle all cases :thinking:

ruiconti committed:
Simplifies and removes redundancy from rule's conditional

ruiconti committed:
Adds more information to the error message

ruiconti committed:
Adds a test scenario that disambigues the need for this specific rule

ruiconti committed:
Add missing conditional guard

ruiconti committed:
Fixes deprecated ubuntu vmImage value

https://github.com/actions/virtual-environments/issues/3287

ruiconti committed:
Merge branch 'master' into rconti/adds-illegal-annonymous-function-error-msg

Gelio commented on azure-pipelines.yml:
```diff
@@ -1,5 +1,5 @@
 pool:
-  vmImage: 'Ubuntu-16.04'
+  vmImage: 'ubuntu-18.04'
```
Thanks for fixing that in #31! I'd appreciate if you rebased this PR on the latest `master` to deduplicate the commit that changes this value

Gelio replied:
I see this change is still in "new" in this PR, but it was already merged to `master`. Could you rebase and drop the commits that are already on `master`?

Gelio commented on src/react-hooks-nesting-walker/react-hooks-nesting-walker.ts:
```diff
@@ -190,6 +190,17 @@ export class ReactHooksNestingWalker extends RuleWalker {
+      /**
+       * Detect if the unnamed expression is wrapped in a illegal function call
+       */
+      if (isIdentifier((ancestor.parent as CallExpression)?.expression)) {
```
I'm afraid we still need to check `isCallExpression(ancestor.parent)`. There may be other AST nodes whose `ancestor.parent.expression` is an identifier. For example, the following snippet:

```ts
obj[(props) => { useRef() }] = 123;
                 ~~~~~~~~ [Hook is in an anonymous function that is passed to an illegal callback. Legal callbacks identifiers that can receive anonymous functions as arguments are memo and forwardRef]
```

which is clearly wrong, because the `ancestor.parent` is an `ElementAccessExpression` (looking at <https://astexplorer.net/>), and its `expression` is the `obj` `Identifier`.

This may be picky, but I would rather show correct error messages even in such peculiar cases :sweat_smile: 

Another test case that IMO shows a wrong error messages with the current rule code:

```ts
new Abc((props) => { useRef() })
                     ~~~~~~~~ [Hook is in an anonymous function that is passed to an illegal callback. Legal callbacks identifiers that can receive anonymous functions as arguments are memo and forwardRef]
```

ruiconti replied:
I may be missing something on this answer but why is the error message wrong in these two examples provided? I believe that they are _correct_, providing that for #1 and #2 we have anonymous functions that are not being passed to the right callback.

Gelio replied:
My reasoning was that the error message in both those cases states, that the hook is used in an anonymous function (correct), which is passed to an illegal callback (incorrect). The constructor or object property access are not callbacks.

Anyway, thanks for adding the check :+1: This results in correct error messages shown in those 2 cases:

```
  obj[(props) => { useRef() }] = 123;
                   ~~~~~~~~           [A hook cannot be used inside of another function]
  new Abc((props) => { useRef() })
                       ~~~~~~~~    [A hook cannot be used inside of another function]
```

Could you add those 2 cases to the tests file, just so we don't regress on this behavior in the future? You can resolve the conversation after that :slightly_smiling_face:

ruiconti replied:
Yup. And thanks for clearing that out, it was indeed the wrong error message.

Gelio commented on test/tslint-rule/anonymous-function-error.ts.lint:
```diff
@@ -0,0 +1,26 @@
+})
+
+const Functor = function() {
+  const cb = React.useCallback(() => {
```
After changing this line to

```diff
-  const cb = React.useCallback(() => {
+  const cb = useCallback(() => {
```

I got the following results from the test

```diff
  const Functor = function() {
    const cb = useCallback(() => {
      const r = React.useRef()
-               ~~~~~~~~~~~~~~ [Hook is in an anonymous function that is passed to an illegal callback. Legal callbacks identifiers that can receive anonymous functions as arguments are memo and forwardRef]
+               ~~~~~~~~~~~~~~ [A hook cannot be used inside of another function]
    }, [])
  }
```

This is because `React.useCallback` is not a `Identifier`, but `useCallback` is, so the condition in the `if` you added is met :worried: 

This leads me to believe that either we need to make the detection logic more robust to detect this case, or (which I still recommend), change the error message of `invalidFunctionExpression` and `invalidFunctionDeclaration` to be generic enough and still describe the problem.

An example error message could be:

> A hook cannot be used inside of a function that is not another hook or a React component. Hooks begin with "use" and React components begin with an uppercase letter

ruiconti replied:
I really believe that we shouldn't conflate these two error messages —invalid function expression and invalid anonymous function callback. Which is the point of this PR. Instead, what if we checked if the ancestor.parent `isComponentOrHookIdentifier`?

Gelio replied:
Looks like it's correct now! Thanks :+1: I'll still do some more thinking about those cases, but we can definitely resolve this conversation

Gelio reviewed (commented):
Funny thing, today is the 3rd anniversary of the release of this TSLint rule :smile: 

Thanks for making further changes! I'm afraid I found a couple of scenarios that result in incorrect error messages being returned :worried: Take a look at the comments

ruiconti commented:
Hey, thanks for the thorough review!

As for the changes of this PR, I believe that ensuring that the ancestor's parent node is not a hook or a component clears out the scenarios you mentioned that were being misclassified. I can't think of any more of them, but I appreciate if you can exhaust this further.

---

After reviewing your comments, I realized that you are correct: the walker is missing being aware of the broader tree path it is in, and whether it is inside a nested valid component or hook. Which seems to be another issue altogether.

Consider the following statement:

```ts
const a = forwardRef(() => {
  while(true) {
    return React.memo(() => {
      const x = useRef();
    })
  }
})
```

~It is an _invalid_ usage of hooks but it doesn't trigger any rules, because current rules rely only on, at most, two levels above `ancestor.parent` in order to trigger some rules~.

Edit: it is a valid snippet  for this tslint rule.

ruiconti committed:
Adds missing check for callback being a valid hook identifier

Gelio commented:
Thanks for introducing more updates :+1: 

> It is an invalid usage of hooks but it doesn't trigger any rules, because current rules rely only on, at most, two levels above ancestor.parent in order to trigger some rules.

In terms of this rule, it does not seem that invalid to me. In the snippet you shared, `useRef` is called within `React.memo`, which is a valid context for a hook call. The fact that we're returning a component type instead of a component instance from `forwardRef` is beyond the responsibility of this rule (and a linter in general, IMO). If you asked me, this is the type of error that should be caught by TypeScript because it's a type error and not strictly related to hooks.

Gelio commented on src/react-hooks-nesting-walker/error-messages.ts:
```diff
@@ -17,4 +17,5 @@ export const ERROR_MESSAGES = {
     'A hook cannot be used inside of another function',
   invalidFunctionExpression: 'A hook cannot be used inside of another function',
   hookAfterEarlyReturn: 'A hook should not appear after a return statement',
+  anonymousFunctionIllegalCallback: `Hook is in an anonymous function that is passed to an illegal callback. Legal callbacks identifiers that can receive anonymous functions as arguments are memo and forwardRef`,
```
Looking at this error message, I am afraid it may be confusing for regular developers who were not working with ASTs/language parsing before.

```suggestion
  anonymousFunctionIllegalCallback: `Hook is used in an anonymous function that is provided as an argument in an unexpected function. Functions that can receive anonymous functions with hook calls are "React.memo" and "React.forwardRef". If this is intended, add a name to the anonymous wrapping function to make it appear as a component.`,
```

Gelio commented on src/react-hooks-nesting-walker/react-hooks-nesting-walker.ts:
```diff
@@ -190,6 +190,21 @@ export class ReactHooksNestingWalker extends RuleWalker {
+       */
+      if (
+        isCallExpression(ancestor.parent) &&
+        isIdentifier(ancestor.parent.expression) &&
+        !isComponentOrHookIdentifier(ancestor.parent.expression)
```
This mostly looks good now :+1: One change that I believe is worth it, is using `isReactComponentDecorator` here. This is because for the following snippet:

```ts
const IllegalComponent = Wrapper(function() {
  useEffect(() => {});
})
```

`Wrapper` is a component identifier, so the code does not report `anonymousFunctionIllegalCallback`, but it reports the more-generic `invalidFunctionExpression` error:

```
  const IllegalComponent = Wrapper(function() {
    useEffect(() => {});
    ~~~~~~~~~~~~~~~~~~~  [A hook cannot be used inside of another function]
  })
```

It should also work in this case, where the current code uses the more-generic message:

```
  const IllegalComponent = React.Wrapper(function() {
    useEffect(() => {});
    ~~~~~~~~~~~~~~~~~~~  [A hook cannot be used inside of another function]
  })
```

I would expect to see the `anonymousFunctionIllegalCallback` message in these cases.

When using 

```suggestion
        !isReactComponentDecorator(ancestor.parent.expression)
```

(note that I removed the `isIdentifier` check, because `isReactComponentDecorator` smartly detects it) we have the following (IMO correct) results in these 2 cases:

```
  const IllegalComponent = Wrapper(function() {
    useEffect(() => {});
    ~~~~~~~~~~~~~~~~~~~  [Hook is used in an anonymous function that is provided as an argument in an unexpected function. Functions that can receive anonymous functions with hook calls are "React.memo" and "React.forwardRef". If this is intended, add a name to the wrapping function to make it appear as a component.]
  })

  const IllegalComponent = React.Wrapper(function() {
    useEffect(() => {});
    ~~~~~~~~~~~~~~~~~~~  [Hook is used in an anonymous function that is provided as an argument in an unexpected function. Functions that can receive anonymous functions with hook calls are "React.memo" and "React.forwardRef". If this is intended, add a name to the wrapping function to make it appear as a component.]
  })
```

However, that also causes changes in other tests:

```diff
    // Usage inside other functions
    [].forEach(() => {
      useEffect(() => {});
-     ~~~~~~~~~~~~~~~~~~~  [A hook cannot be used inside of another function]
+     ~~~~~~~~~~~~~~~~~~~  [Hook is used in an anonymous function that is provided as an argument in an unexpected function. Functions that can receive anonymous functions with hook calls are "React.memo" and "React.forwardRef". If this is intended, add a name to the wrapping function to make it appear as a component.]
    })
```

I believe that is acceptable, but not sure if it's worth it - the new error message (either the one you suggested or the one from me) can mislead the engineer.

Overall, it seems like a hard problem to detect function calls and differentiate between situations when the user is using a hook in a component-like function, or a regular callback (e.g. `array.forEach`'s callback) without [an equivalent of `isSomewhereInsideComponentOrHook`](https://github.com/Gelio/tslint-react-hooks/pull/30#discussion_r798359325).

I'm not sure how you want to proceed here. The current implementation adds this new type of error only in very special scenarios (only when the called function's name does not look like a hook/component, so uppercase function names are ignored) and does not detect all cases, even though the implementation suggests it does. We could modify the implementation to say that it detects **some** invalid function calls, or we could implement `isSomewhereInsideComponentOrHook` and make it more robust.

Personally, I would rather implement a more robust solution to try to make the error messages more precise

ruiconti commented:
> In the snippet you shared, useRef is called within React.memo, which is a valid context for a hook call. 

That is correct. I was attempting to exercise corner cases that relied on a broader context awareness, but it's not the case for this particular linter —which relies mostly on the enclosing scope in which the hook is defined. Appreciate your patience on clearing those out 👍

Please let me know if you think there's anything left prior to sign off.

ruiconti committed:
Adds test scenario to cover for potential amibguous matching

Hooks defined within anonymous functions which are not being passed to a callback. Which makes them automatically illegal.

Gelio reviewed (commented):
Sorry for the long wait, I have bit a bit busy last week. Also, I have just noticed some of my comments were not published during my last review. I'm posting them now

I will have time to review more changes next weekend.
